share/doc/homebrew/Node-for-Formula-Authors.md

```diff
@@ -0,0 +1,115 @@
+
+# Dependencies
+
+Node modules, which are compatible with the latest Node version should declare a dependencies on the `node` formula.
+
+```ruby
+depends_on "node"
+```
+
+If your formula requires to be executed with an older Node version you must vendor this older Node version as done in the [`kibana` formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/kibana.rb).
+
+### Special requirements for native addons
+
+If your node module is a native addon or has a native addon somewhere in it's dependency tree you have to declare an additional dependency. Since the compilation of the native addon results in a invocation of `node-gyp` we need an additional build time dependency on `:python` (because gyp depends on Python 2.7).
+
+```ruby
+depends_on :python => :build
+```
+
+Please also note, that such a formula would only be compatible with the same node major version it originally was compiled with. This means that we need to revision every formula with a node native addon with every major version bump of the `node` formula. To make sure we don't overlook your formula on a node major version bump, write a meaningful test which would fail in such a case (invoked with an ABI incompatible node version).
+
+# Installation
+
+Node modules should be installed to `libexec`. This prevents the node modules from contaminating the global `node_modules`, which is important so that npm doesn't try to manage Homebrew-installed node modules.
+
+In the following we distinguish between 2 type of node module using formulas:
+* formulas for standard node modules compatible with npm's global module format which should use [`std_npm_install_args`](#installing-global-style-modules-with-std_npm_install_args-to-libexec) (like [`azure-cli`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/azure-cli.rb) or [`autocode`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/autocode.rb)) and
+* formulas were the `npm install` step is only one of multiple not exclusively node related install steps (not compatible with npm's global module format) which have to use [`local_npm_install_args`](#installing-module-dependencies-locally-with-local_npm_install_args) (like [`elixirscript`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/elixirscript.rb) or [`kibana`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/kibana.rb))
+
+Both methods have in common, that they are setting the correct environment for using npm inside Homebrew up and returning the arguments for invoking `npm install` for their specific use cases. This includes fixing an important edge case with the npm cache (caused by the way Homebrew patches the `$HOME` environment variable) by using our own custom `npm_cache` inside `HOMEBREW_CACHE`, which would otherwise result in very long build times and high disk space usage.
```

Comment:
It'd be nice to be a little more specific on our hostage-taking of `$HOME` so there's no implication it's a broad thing. Maybe something like:

```
(Caused by Homebrew's redirection of `$HOME` during the build and test process)
```
or such?